### PR TITLE
Retrieve & show file via intent

### LINKED ISSUE #3933: Open file/folder from Talk
I want to click on a file preview in Talk and open a file or folder inside Nextcloud Files. Current approach is as follows:

- [x] Check if we have the app 
- [x]  Check if Nc app has the correct account
- [ ]  Check if Nc app is the right version

The way I do this is as follows and is brandable + it supports various (similar) apps installed since it doesn't hook into the "nc://" mechanism.

```
                    Intent filesAppIntent = new Intent(Intent.ACTION_VIEW, null);
                    final ComponentName cn = new ComponentName(context.getString(R.string.nc_import_accounts_from), "com.owncloud.android.ui.activity.FileDisplayActivity");
                    filesAppIntent.setComponent(cn);
                    filesAppIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    filesAppIntent.setPackage(context.getString(R.string.nc_import_accounts_from));
                    Bundle options = new Bundle();
                    options.putString(BundleKeys.KEY_ACCOUNT, accountString);
                    options.putString(BundleKeys.KEY_FILE_PATH, "/" + message.getSelectedIndividualHashMap().get("path");
                    context.startActivity(filesAppIntent, options);

```
nc_import_accounts_from - the package name of the files app
BundleKeys.KEY_ACCOUNT - KEY_ACCOUNT
BundleKeys.KEY_FILE_PATH - KEY_FILE_PATH

Let me know if you have any questions @nickvergessen, @tobiasKaminsky and @AndyScherzinger 

If this is sound, please implement handling of the intent inside the Files app and release a version so I can include a version check inside the Talk app.

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
Fix #3933 

@mario I tested it with app "intents" and used this: 
- action.view
- app: com.nextcloud.client
- component: com.owncloud.android.ui.activity.FileDisplayActivity
- KEY_ACCOUNT, e.g. tobi@10.0.2.2/nc
- KEY_FILE_ID: e.g. 81234

This works 
- when app is force closed
- when app is running, but in background
- retrieving unknown file/folder
- showing a loading dialog
- scroll to file & highlight it (grey background)

Can I test this with a talk version?

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

### CODE REVIEW

**Comment by mario** [comment]
@tobiasKaminsky there's a branch called "smaller-improvements" in Talk where I've done a small change to allow you to test this. Once it's working, let me know what version code will this be released under for the Files app - thanks!

**Comment by mario** [comment]
Btw. I think the activity needs to be exported so it could be started from another app?

**Comment by tobiasKaminsky** [comment]
> Btw. I think the activity needs to be exported so it could be started from another app?

> The default value is false for activity, receiver, and service components that do not specify any intent filters; it is true for activity, receiver, and service components that do have intent filters (implying they expect to be invoked by others who do not know their particular component name) and for all content providers.

So this is done automatically by Android.

**Comment by tobiasKaminsky** [comment]
I had to change the way of setting extras:
https://github.com/nextcloud/talk-android/commit/94689d9cb9981c1004649e65f34fe407ac26e343

Now it works :tada:

**Comment by mario** [comment]
👍 

Nice. I'll try to review soon. How do you handle if the current account in files app is different from the one I send?

**Comment by tobiasKaminsky** [comment]
> How do you handle if the current account in files app is different from the one I send?

I'll just switch the account.

**Comment by mario** [comment]
Given that I want to release this soon, do you know what version number will the Files app be when you release this? (e.g. we need this soon :P)

**Comment by tobiasKaminsky** [comment]
Is 3.6.1 end of next week fast enough?
Then I can do RC1 on monday.

**Comment by mario** [comment]
@tobiasKaminsky it is, I just need the versionCode of the RC1 that you'll use with this PR merged, so I can do a Talk release before I leave for a vacation :)

**Comment by tobiasKaminsky** [comment]
Code: 30060151
Name: 3.6.1 RC1

File: src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 1]
```diff
@@ -2595,4 +2612,101 @@ public void setSearchQuery(String query) {
...
+
+                    if (remoteOperationResult.isSuccess() && remoteOperationResult.getData() != null) {
+                        String remotePath = ((RemoteFile) remoteOperationResult.getData().get(0)).getRemotePath();
+
+                        ReadFileRemoteOperation operation = new ReadFileRemoteOperation(remotePath);
+                        RemoteOperationResult result = operation.execute(getAccount(), context);
+
+                        if (!result.isSuccess()) {
+                            Exception exception = result.getException();
+                            String message = "Error during saving file with parents: " + remotePath + " / "
```
Comment (mario) [inline src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java]: I guess the error message is ... wrong? Or just strange?

File: src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java [hunk 2]
```diff
@@ -2595,4 +2612,101 @@ public void setSearchQuery(String query) {
...
 
+    private void handleOpenFileViaIntent(Intent intent) {
+        showLoadingDialog("Retrieving file…");
+
+        String accountName = intent.getStringExtra("KEY_ACCOUNT");
+        setAccount(getUserAccountManager().getAccountByName(accountName));
+
+        String fileId = String.valueOf(intent.getStringExtra("KEY_FILE_ID"));
+
+        if (!"null".equals(fileId)) {
```
Comment (mario) [inline src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Hm, this should never be "null"? :)

File: src/main/java/com/owncloud/android/ui/adapter/OCFileListAdapter.java [hunk 3]
```diff
@@ -128,6 +130,7 @@
     private List<ThumbnailsCacheManager.ThumbnailGenerationTask> asyncTasks = new ArrayList<>();
     private boolean onlyOnDevice;
     private boolean showShareAvatar;
+    @Getter @Setter private OCFile highlightedItem;
```
Comment (mario) [inline src/main/java/com/owncloud/android/ui/adapter/OCFileListAdapter.java]: Do you need a public getter and setter here or?

**Comment by mario** [comment]
Few questions.

File: src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java (hunk 1 above)
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java]: Nice catch, was c&p ;-)

File: src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java (hunk 2 above)
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/activity/FileDisplayActivity.java]: As this is triggered by every intent with action_view, I just wanted to have it bulletproofed.

File: src/main/java/com/owncloud/android/ui/adapter/OCFileListAdapter.java (hunk 3 above)
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/adapter/OCFileListAdapter.java]: Right, a getter is indeed not needed.

**Comment by tobiasKaminsky** [comment]
I extracted AsyncTask to be in a separate class.
Can you review and test again, and then I can do RC1 today.

**Comment by mario** [comment]
On vacation, but looks decent.

**Comment by AndyScherzinger** [comment]
fine with me code-wise

**Pull request merged**